modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTable.java

```diff
@@ -0,0 +1,203 @@
+import io.siddhi.core.event.stream.StreamEvent;
+import io.siddhi.core.event.stream.StreamEventCloner;
+import io.siddhi.core.event.stream.StreamEventFactory;
+import io.siddhi.core.executor.VariableExpressionExecutor;
+import io.siddhi.core.table.record.RecordTableHandler;
+import io.siddhi.core.util.collection.AddingStreamEventExtractor;
+import io.siddhi.core.util.collection.operator.CompiledCondition;
+import io.siddhi.core.util.collection.operator.MatchingMetaInfoHolder;
+import io.siddhi.core.util.collection.operator.Operator;
+import io.siddhi.core.util.config.ConfigReader;
+import io.siddhi.query.api.definition.Attribute;
+import io.siddhi.query.api.definition.TableDefinition;
+import io.siddhi.query.api.expression.Expression;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * common interface for FIFO, LRU, and LFU cache tables
+ */
+public abstract class CacheTable extends InMemoryTable {
+    private int maxSize;
+    boolean cacheExpiryEnabled;
+    SiddhiAppContext siddhiAppContext;
+
+    public void initCacheTable(TableDefinition cacheTableDefinition, ConfigReader configReader,
+                               SiddhiAppContext siddhiAppContext, RecordTableHandler recordTableHandler,
+                               boolean cacheExpiryEnabled, int maxSize, String cachePolicy) {
+        this.maxSize = maxSize;
+        this.cacheExpiryEnabled = cacheExpiryEnabled;
+        this.siddhiAppContext = siddhiAppContext;
+        addRequiredFieldsToCacheTableDefinition(cacheTableDefinition, cacheExpiryEnabled);
+
+        // initialize cache table
+        MetaStreamEvent cacheTableMetaStreamEvent = new MetaStreamEvent();
+        cacheTableMetaStreamEvent.addInputDefinition(cacheTableDefinition);
+        for (Attribute attribute : cacheTableDefinition.getAttributeList()) {
+            cacheTableMetaStreamEvent.addOutputData(attribute);
+        }
+
+        StreamEventFactory cacheTableStreamEventFactory = new StreamEventFactory(cacheTableMetaStreamEvent);
+        StreamEventCloner cacheTableStreamEventCloner = new StreamEventCloner(cacheTableMetaStreamEvent,
+                cacheTableStreamEventFactory);
+        super.initTable(cacheTableDefinition, cacheTableStreamEventFactory,
+                cacheTableStreamEventCloner, configReader, siddhiAppContext, recordTableHandler);
+    }
+
+    public void addStreamEventUptoMaxSize(StreamEvent streamEvent) {
+        int sizeAfterAdding = 0;
+        ComplexEventChunk<StreamEvent> addEventsLimitCopy = new ComplexEventChunk<>(true);
+        StreamEvent streamEventIter;
+        while (true) {
+            sizeAfterAdding++;
+            streamEventIter = streamEvent.getNext();
+            if (sizeAfterAdding == maxSize || streamEventIter.getNext() == null) {
+                streamEventIter.setNext(null);
+                break;
+            }
+        }
+        addEventsLimitCopy.add(streamEvent);
```

Comment:
Since you already have the first and the last events use 
public ComplexEventChunk(E first, E last, boolean isBatch) ; constructor

Reply:
changed the function slightly so now it is not needed